### ISSUE TITLE
why onnx model is twice bigger than pt model.

### ISSUE THREAD

**Issue by tengshaofeng** [ultralytics/ultralytics#17464]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I transfer model from format of pt to format of onnx. The model is  twice bigger than pt model.
"from ultralytics import YOLO
model = YOLO("yolo11x.pt")
path = model.export(format="onnx", half=True)
"

### Additional

![image](https://github.com/user-attachments/assets/c96d5e6d-c19e-4adc-be8c-c5901614077e)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tengshaofeng, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com) for guidance on using our models and deployment formats, where you might find insights related to differences in model sizes between formats.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us pinpoint the issue. Your snippet is a great start! Consider adding specific details about your environment or any anomalies you encountered during the conversion process.

Join the Ultralytics community where it suits you best. For real-time chat, hop onto [Discord](https://ultralytics.com/discord) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com). You can also explore our [Subreddit](https://reddit.com/r/ultralytics) to engage with other users.

## Upgrade

Make sure you have the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Experiencing this issue in different environments could help identify its consistency. YOLO can be tested in these setups:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. These tests validate YOLO's operations across different platforms.

This is an automated response, but rest assured, an Ultralytics engineer will follow up with you shortly to provide further assistance!

**Comment by Y-T-G** [comment]
`half=True` doesn't work unless you also pass CUDA device `device=0`.

**Comment by Y-T-G** [comment]
It should said that in a warning too

**Comment by Y-T-G** [comment]
It should have mentioned that in a warning too

**Comment by tengshaofeng** [comment]
> `half=True` doesn't work unless you also pass CUDA device `device=0`.

thanks  for your reply. I have no gpu. I just do it on my macbook.  And do "path = model.export(format="onnx", half=True)" without seting device=0

**Comment by Y-T-G** [comment]
Without GPU, ONNX export only uses FP32. So it's twice the size. Because the original weights are FP16.

**Comment by tengshaofeng** [comment]
> Without GPU, ONNX export only uses FP32. So it's twice the size. Because the original weights are FP16.

thanks for your reply. I test on yolov10 and yolo11 using the same export code. It is strange that, yolov10.onnx size is almost same with yolov10.pt, but  yolo11.onnx size is twice bigger than  yolov10.pt.
![image](https://github.com/user-attachments/assets/c6de49a3-0d87-4c28-9af3-525788d8bcd7)

**Comment by Y-T-G** [comment]
Because the pretrained weights for YOLOv10 are in FP32.

**Comment by tengshaofeng** [comment]
> Because the pretrained weights for YOLOv10 are in FP32.

if pretrained weights for YOLOv10 are in FP32 with 128M, then after  model.export(format="onnx", half=True), it should be 128/2=64M

**Comment by Y-T-G** [comment]
Like I said, ONNX half export only works if you have GPU and you pass `device=0`

**Comment by Y-T-G** [comment]
Also where did you get the YOLOv10x weight from? Because the weights in `ultralytics` is 61.4MB, which is because it's in FP16. The one in your screenshot is in FP32, so it's double.

```python
In [3]: model = YOLO("yolov10x.pt")
Downloading https://github.com/ultralytics/assets/releases/download/v8.3.0/yolov10x.pt to 'yolov10x.pt'...
100%|█████████████████████████████| 61.4M/61.4M [00:00<00:00, 97.6MB/s]
```

**Comment by Y-T-G** [comment]
You're probably using an old version of `ultralytics` which had FP32 weights for YOLOv10. In the latest version, the weights are FP16.

**Comment by tengshaofeng** [comment]
> You're probably using an old version of `ultralytics` which had FP32 weights for YOLOv10. In the latest version, the weights are FP16.

understand that, thanks so much.

**Comment by inisis** [comment]
@Y-T-G  I tried this,
```
        torch.onnx.export(
            self.model.half().cpu(),  # dynamic=True only compatible with cpu
            self.im.half().cpu(),
            f,
            verbose=False,
            opset_version=opset_version,
            do_constant_folding=True,  # WARNING: DNN inference with torch>=1.12 may require do_constant_folding=False
            input_names=["images"],
            output_names=output_names,
            dynamic_axes=dynamic or None,
        )
```

it seesm to work but takes too much time
```
Ultralytics 8.3.28 🚀 Python-3.10.15 torch-2.5.1+cpu CPU (Intel Xeon Platinum 8378C 2.80GHz)
WARNING ⚠️ half=True only compatible with GPU export, i.e. use device=0
YOLO11x summary (fused): 464 layers, 56,919,424 parameters, 0 gradients, 194.9 GFLOPs

PyTorch: starting from 'yolo11x.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (109.3 MB)

ONNX: starting export with onnx 1.17.0 opset 19...

ONNX: slimming with onnxslim 0.1.35...
ONNX: export success ✅ 203.9s, saved as 'yolo11x.onnx' (108.8 MB)

Export complete (205.4s)
Results saved to /root
Predict:         yolo predict task=detect model=yolo11x.onnx imgsz=640  
Validate:        yolo val task=detect model=yolo11x.onnx imgsz=640 data=/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app
```

**Comment by Y-T-G** [comment]
@inisis think FP16 inference is slow on CPU which is why it's disabled.

**Comment by Y-T-G** [comment]
@inisis This only keeps the weights in FP16? The inference could still use FP32?

**Comment by inisis** [comment]
No, inputs must be converted to fp16, onnxslim uses onnxconverter_common to convert float32 weights to half precision, and onnxslim can also convert half precision to full precision, details can be found here, and I'm also considering add bfloat16 support.
https://github.com/inisis/OnnxSlim/blob/main/onnxslim/core/__init__.py#L162-L186

**Comment by Y-T-G** [comment]
@inisis It would be nice if the ONNX CPU users can somehow still use FP32 inference while the ONNX model weights are in FP16 for smaller file size.

> onnxslim can also convert half precision to full precision, details can be found here, and I'm also considering add bfloat16 support.

So this should be able to convert the FP16 weights to FP32 for CPU users before inference, right?

**Comment by inisis** [comment]
@Y-T-G onnxslim can convert half precision to full precision without losing accuracy, but the bottleneck lies in converting half precision model to onnx, we do can convert full precision pytorch model to onnx then cast onnx down to half precision, but this will lose some accuracy, if accuracy does not matter, we can give it a try.

**Comment by Y-T-G** [comment]
@inisis Full precision weights actually doesn't matter for `ultralytics` because `ultralytics` doesn't use full precision weights by default even for PyTorch weights.

**Comment by inisis** [comment]
@Y-T-G so why users have to use full precision for inference, we can add a cast op after inputs in onnx model to make sure half precision inference

**Comment by Y-T-G** [comment]
@inisis The inference in full precision can still have different scores from half precision because the intermediate calculations can use full precision. But the PyTorch weights themselves are stored as half precision and converted to full precision during training or inference.

**Comment by inisis** [comment]
Onnx backend can do it too, and onnx slim can help!

**Comment by Y-T-G** [comment]
@inisis Great! You can try opening a PR.

**Comment by inisis** [comment]
https://github.com/ultralytics/ultralytics/pull/17498 @Y-T-G please review it
